### PR TITLE
Implement routing for Linode detail page

### LINKED ISSUE #86: Implement routing for Linode detail page

### PR DESCRIPTION
**Pull request by SirCmpwn**
Note that the components for each tab need to be added in
src/linodes/index.js now instead of directly in the detail page.

Closes #86.

### CODE REVIEW

**Comment by SirCmpwn** [comment]
This involves some hacks to get routing to make sense in the context of react-tabs.

File: scss/components/card.scss [hunk 1]
```diff
@@ -24,6 +24,11 @@
                 background: white;
                 border-bottom: 0;
             }
+
+            a {
+                color: inherit;
+                text-decoration: inherit;
```
Comment (eatonphil) [inline scss/components/card.scss]: These look like defaults.

File: src/linodes/index.js [hunk 2]
```diff
@@ -1,14 +1,34 @@
...
 
 import IndexPage from './layouts/IndexPage';
 import LinodeDetailPage from './layouts/LinodeDetailPage';
 import CreateLinodePage from './layouts/CreateLinodePage';
 
+/* eslint-disable react/prop-types */
+class Placeholder extends Component {
+  constructor() {
+    super();
+    this.render = this.render.bind(this);
```
Comment (eatonphil) [inline src/linodes/index.js]: Huh? You are binding a class method? Why is this necessary?

File: scss/components/card.scss (hunk 1 above)
Comment (SirCmpwn) [inline scss/components/card.scss]: It's to prevent the links from being rendered as links, but to instead look like tabs.

Comment (SirCmpwn) [inline scss/components/card.scss]: Before



After

File: src/linodes/index.js (hunk 2 above)
Comment (SirCmpwn) [inline src/linodes/index.js]: ...this is the exact same pattern as was used in the previous discussion.

**Comment by SirCmpwn** [comment]
Coverage seems to have been reduced because I added (and did not test) the placeholder code in index.js. Don't care.

File: src/linodes/layouts/LinodeDetailPage.js [hunk 3]
```diff
@@ -147,42 +148,46 @@ export class LinodeDetailPage extends Component {
...
+
+    const pathname = location ? location.pathname : tabList[0].link;
+
+    let selected = tabList[0];
+    for (let i = 1; i < tabList.length; ++i) {
+      if (pathname === tabList[i].link) {
+        selected = tabList[i];
+        break;
+      }
+    }
```
Comment (eatonphil) [inline src/linodes/layouts/LinodeDetailPage.js]: Or:

    const selected = tabList.reduce((lastTab, thisTab) => pathname === thisTab.link ? thisTab : lastTab);

Comment (SirCmpwn) [inline src/linodes/layouts/LinodeDetailPage.js]: Sure.

File: src/linodes/index.js (hunk 2 above)
Comment (eatonphil) [inline src/linodes/index.js]: The only reason you ever need to do this is if you plan on calling `this.render`.

Comment (SirCmpwn) [inline src/linodes/index.js]: No, you also need it if `render` uses `this`. Which it does.

Comment (eatonphil) [inline src/linodes/index.js]: So you're saying the code does not work without this line?

Comment (SirCmpwn) [inline src/linodes/index.js]: It does work. I guess React binds this for you when it invokes render. Pushed a change.

Comment (eatonphil) [inline src/linodes/index.js]: I think you may be misunderstanding how `this` works. The reason you `bind` this in certain contexts is because a function will inherit `this` from the `function` (not `() => {}`) that calls it. Javascript would SUCK if you always had to be binding this to ever access it at all. Creating a function in a class implicitly binds `this`. Again, the only reason you explicitly do it is when it's being called by some other function whose `this` is not equal to the `this` you want.

Comment (SirCmpwn) [inline src/linodes/index.js]: I understand how it works.

**Comment by SirCmpwn** [comment]
Checks are green. Merge?

**Comment by eatonphil** [comment]
Looks good.

**Comment by SirCmpwn** [comment]
Thanks!

**Comment by eatonphil** [comment]
My explanation was wrong about `this`. So cleaning that up for my own sake.

`this` is bound to be the object who parents the function. Calling `bind` overwrites that. It is used in React when you are passing a callback (where `this` within the callback function is intended to be the parent) from a parent to a child. Calling the callback requires this literal `this.callback()` that would assign `this` within the callback to be the child if `this` were not bound in the parent level.

Since `render` is never called from any other parent except the class it exists in, `this` will never be bound to anything other than what you'd expect. That's why it is not necessary in this case (or ever, probably).
